### PR TITLE
Remove dependency between command line flags and providers

### PR DESCRIPTION
**Pull request by bogdandrutu**
Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

### CODE REVIEW

File: service/parserprovider/properties.go [hunk 1]
```diff
@@ -26,28 +25,30 @@ import (
...
-// NewSetFlag returns a ParserProvider, that provides a config.Map from set flag(s) properties.
+// NewPropertiesProvider returns a ParserProvider, that provides a config.Map from set flag(s) properties.
 //
-// The implementation reads set flag(s) from the cmd and concatenates them as a "properties" file.
-// Then the properties file is read and properties are set to the loaded Parser.
-func NewSetFlag() ParserProvider {
-	return &setFlagProvider{}
+// The implementation parses the properties and concatenates them as a "properties" file.
+// Then the properties file is read and properties are set to the returned config.Map.
+func NewPropertiesProvider(props []string) ParserProvider {
```
Comment (mx-psi) [inline service/parserprovider/properties.go:36]: Adding a comment here about the `props` argument format would help (something like "key-value list separated by equal sign with a `.` as key delimiter" maybe?)

File: service/parserprovider/file.go [hunk 2]
```diff
@@ -22,28 +22,31 @@ import (
...
 )
 
-type fileProvider struct{}
+type fileProvider struct {
+	fileName string
+}
 
-// NewFile returns a new ParserProvider that reads the configuration from a file configured
+// NewFileProvider returns a new ParserProvider that reads the configuration from a file configured
 // via the --config command line flag.
```
Comment (mx-psi) [inline service/parserprovider/file.go:30]: It's more general now
```suggestion
```

**Review by mx-psi — COMMENTED** [review]
The approach makes sense to me, I left two small comments

File: service/collector_windows.go [hunk 3]
```diff
@@ -90,8 +98,7 @@ func (s *WindowsService) start(elog *eventlog.Log, colErrorChannel chan error) e
 	// col.Run blocks until receiving a SIGTERM signal, so needs to be started
 	// asynchronously, but it will exit early if an error occurs on startup
 	go func() {
-		cmd := NewCommand(s.col)
-		colErrorChannel <- cmd.Execute()
+		colErrorChannel <- col.Run(context.Background())
```
Comment (mx-psi) [inline service/collector_windows.go:101]:
```suggestion
		colErrorChannel <- s.col.Run(context.Background())
```

File: service/parserprovider/file.go (hunk 2 above)
Comment (bogdandrutu) [reply service/parserprovider/file.go:30]: Done.

File: service/collector_windows.go (hunk 3 above)
Comment (bogdandrutu) [reply service/collector_windows.go:101]: Done.

File: service/parserprovider/properties.go (hunk 1 above)
Comment (bogdandrutu) [reply service/parserprovider/properties.go:36]: Done.

**Review by mx-psi — APPROVED** [review]
LGTM

File: service/parserprovider/properties.go [hunk 4]
```diff
@@ -26,28 +25,31 @@ import (
...
+}
 
-// NewPropertiesMapProvider returns a MapProvider, that provides a config.Map from set flag(s) properties.
+// NewPropertiesMapProvider returns a MapProvider, that provides a config.Map from the given properties.
 //
-// The implementation reads --set flag(s) from the cmd and concatenates them as a "properties" file.
-// Then the properties file is read and properties are set to the loaded Parser.
-func NewPropertiesMapProvider() MapProvider {
-	return &setFlagMapProvider{}
+// Properties must follow the Java properties format, key-value list separated by equal sign with a "."
```
Comment (codeboten) [inline service/parserprovider/properties.go:34]: Why follow java properties format here? I'm not against it just curious

File: service/parserprovider/default.go [hunk 5]
```diff
@@ -14,9 +14,11 @@
...
 package parserprovider
 
-// NewDefaultMapProvider is the default MapProvider and it creates configuration from a file
-// defined by the --config command line flag and overwrites properties from --set
-// command line flag (if the flag is present).
-func NewDefaultMapProvider() MapProvider {
-	return NewExpandMapProvider(NewMergeMapProvider(NewFileMapProvider(), NewPropertiesMapProvider()))
+// NewDefaultMapProvider returns the default MapProvider, and it creates configuration from a file
+// defined by the given configFile and overwrites fields from the properties.
+func NewDefaultMapProvider(configFile string, properties []string) MapProvider {
```
Comment (codeboten) [inline service/parserprovider/default.go:19]: i would rename configFile configFileName

File: service/parserprovider/default.go [hunk 6]
```diff
@@ -14,9 +14,11 @@
 
 package parserprovider
 
-// NewDefaultMapProvider is the default MapProvider and it creates configuration from a file
-// defined by the --config command line flag and overwrites properties from --set
-// command line flag (if the flag is present).
-func NewDefaultMapProvider() MapProvider {
-	return NewExpandMapProvider(NewMergeMapProvider(NewFileMapProvider(), NewPropertiesMapProvider()))
+// NewDefaultMapProvider returns the default MapProvider, and it creates configuration from a file
+// defined by the given configFile and overwrites fields from the properties.
```
Comment (codeboten) [inline service/parserprovider/default.go:18]:
```suggestion
// defined by the given configFile and overwrites fields using properties.
```

**Review by codeboten — APPROVED** [review]
just a couple of non-blocking suggestions

File: service/parserprovider/properties.go (hunk 4 above)
Comment (bogdandrutu) [reply service/parserprovider/properties.go:34]: This was already the format, I just clarified.

File: service/parserprovider/default.go (hunk 5 above)
Comment (bogdandrutu) [reply service/parserprovider/default.go]: Done.

**Pull request merged**